service/awsconfig/v10/ebs/ebs.go

```diff
@@ -0,0 +1,191 @@
+package ebs
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/cenkalti/backoff"
+	"github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+
+	"github.com/giantswarm/aws-operator/service/awsconfig/v10/key"
+)
+
+const (
+	cloudProviderClusterTagValue        = "owned"
+	cloudProviderPersistentVolumeTagKey = "kubernetes.io/created-for/pv/name"
+	nameTagKey                          = "Name"
+)
+
+type Config struct {
+	Client EC2Client
+	Logger micrologger.Logger
+}
+
+type EBS struct {
+	client EC2Client
+	logger micrologger.Logger
+}
+
+func New(config Config) (*EBS, error) {
+	if config.Client == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Client must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	c := &EBS{
+		client: config.Client,
+		logger: config.Logger,
+	}
+
+	return c, nil
+}
+
+// DeleteVolume deletes an EBS volume with retry logic.
+func (e *EBS) DeleteVolume(ctx context.Context, volumeID string) error {
+	e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting ebs volume %s", volumeID))
+
+	deleteOperation := func() error {
+		_, err := e.client.DeleteVolume(&ec2.DeleteVolumeInput{
+			VolumeId: aws.String(volumeID),
+		})
+		if IsVolumeNotFound(err) {
+			// Fall through.
+			return nil
+		}
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		return nil
+	}
+	deleteNotify := func(err error, delay time.Duration) {
+		e.logger.LogCtx(ctx, "level", "error", fmt.Sprintf("deleting ebs volume failed, retrying with delay %.0fm%.0fs: '%#v'", delay.Minutes(), delay.Seconds(), err))
```

Comment:
AFAIK this should panic because the "keyvals" are uneven. The `message` key is missing. I would also like to follow operatorkit and put the `%#v` format for the actual error via the `stack` key. The message is fine but the error/stack itself should be separate. This makes EFK more powerful and humans can observe the log line better.

Reply:
Thanks it didn't panic but there were some `(MISSING)` entries in the logs. Done and much nicer to use the stack key.